v0/destinations/kafka/transform.js

```diff
@@ -1,36 +1,83 @@
+/* eslint-disable no-restricted-syntax */
+const groupBy = require("lodash/groupBy");
+const cloneDeep = require("lodash/cloneDeep");
+const { getDynamicMeta } = require("../../../adapters/utils/networkUtils");
 const { getIntegrationsObj } = require("../../util");
+const { TRANSFORMER_METRIC } = require("../../util/constant");
+const ErrorBuilder = require("../../util/error");
 
 function batch(destEvents) {
   const respList = [];
-  const batchedRequest = [];
-  const metadata = [];
-  destEvents.forEach(event => {
-    metadata.push(event.metadata);
-    batchedRequest.push(event.message);
-  });
-  respList.push({
-    batchedRequest,
-    metadata,
-    destination: destEvents[0].destination
+
+  // Grouping the events by topic
+  const groupedEvents = groupBy(destEvents, event => {
+    return event.message.topic;
   });
 
+  // Creating a batched request for each topic
+  // we are grouping the events based on topics
+  // and creating a batched request for each topic
+  // example: input events = [{event1,topic1},{event2,topic1},{event3,topic2}]
+  // out from transformer:  {batchedRequest:[{event1},{event2}]}, {batchedRequest:[{event3}]} (2 multilexed responses)
+  for (const events of Object.values(groupedEvents)) {
+    const response = {};
+    response.batchedRequest = events.map(event => event.message);
+    response.metadata = events.map(event => event.metadata);
+    response.destination = events[0].destination;
+    respList.push(response);
+  }
+
```

Comment:
If respList has size as 2, will server store these 2 as different jobId's for router ?

Reply:
So it will be handled like this
Resp1 = [e1,e2,e3], Resp2=[e4,e5]
No these each events metadata will tracked in rt table (jobid is a property in metadata)

in produce if batching is enabled resp1 will be processed if it fails [e1,e2,e3] all jobs fail

Reply:
Ok each event will have jobId in rt_jobs table. But during produce, Resp1 & Resp2 will be treated as two separate requests for delivery.

Reply:
so even in case of batching we won't be creating new jobID in rt_jobs right?

Reply:
Correct